xdsl/builder.py

```diff
@@ -107,6 +127,113 @@ def region(
         else:
             return Builder._region_args(input)
 
+    @staticmethod
+    def _implicit_region_no_args(func: Callable[[], None]) -> Region:
+        """
+        Generates a single-block region.
+        """
+        block = Block()
+        builder = Builder(block)
+
+        with Builder._push_implicit_builder(builder):
+            func()
+
+        return Region.from_block_list([block])
+
+    @staticmethod
+    def _implicit_region_args(
+        input_types: list[Attribute] | ArrayAttr[Attribute]
+    ) -> Callable[[_CallableImplicitRegionFuncType], Region]:
+        """
+        Decorator for constructing a single-block region, containing the implementation of a
+        function with some input arguments.
+        """
+
+        if isinstance(input_types, ArrayAttr):
+            input_type_seq = input_types.data
+        else:
+            input_type_seq = input_types
+
+        def wrapper(func: _CallableImplicitRegionFuncType) -> Region:
+            block = Block.from_arg_types(input_type_seq)
+            builder = Builder(block)
+
+            with Builder._push_implicit_builder(builder):
+                func(block.args)
+
+            region = Region.from_block_list([block])
+            return region
+
+        return wrapper
+
+    @overload
+    @staticmethod
+    def implicit_region(
+        input: list[Attribute] | ArrayAttr[Attribute]
+    ) -> Callable[[_CallableImplicitRegionFuncType], Region]:
+        """
+        Annotation used to construct a Region tuple from a function.
+        The annotation can be used in two ways:
+
+        For regions that have inputs or outputs:
+        ```
+        @Builder.implicit_region(input_types)
+        def func(args: tuple[BlockArgument, ...]) -> None:
+            ...
+        ```
+
+        For regions that don't have inputs or outputs:
+        ``` python
+        @Builder.implicit_region
+        def func(builder: Builder) -> None:
+            ...
+        ```
+        """
+        ...
+
+    @overload
+    @staticmethod
+    def implicit_region(input: Callable[[], None]) -> Region:
+        ...
+
+    @staticmethod
+    def implicit_region(
+        input: list[Attribute] | ArrayAttr[Attribute]
+        | Callable[[], None]
+    ) -> Callable[[_CallableImplicitRegionFuncType], Region] | Region:
+        if isinstance(input, Callable):
+            return Builder._implicit_region_no_args(input)
+        else:
+            return Builder._implicit_region_args(input)
+
+    @classmethod
+    def _push_implicit_builder(cls, builder: Builder) -> _ImplicitBuilder:
+        return _ImplicitBuilder(cls._local.bb, builder)
+
+    @classmethod
+    def get_implicit_builder(cls) -> Builder | None:
```

Comment:
Maybe we should rename it to `get_nearest_implicit_builder`?
Also, do we have a use case of keeping this public? I feel we should maybe make this private, unless you have other ideas in mind